### PR TITLE
[CoreWorker] Cache the named actors in core worker locally.

### PR DESCRIPTION
**Pull request by jovany-wang**
## Why are these changes needed?
It's inefficient to fetch named actor from GCS every time since there aren't many named actors in most user cases, but every named actor is used frequently.
The bulk of users use the `get_named_actor` API directly(without any storage for the named actors) :
```python
ray.get_actor("xxx")      // every where
```

In Java:
```java
Ray.getActor("xxxx");      // every where
```

- Move the code of fetching actor info to actor manager.
- Add delete callback to remove the cache.
- Add C++ test cases to cover cache and evict.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simon-mo** [comment]
This is great! Can you handle the deletion case as well? In Ray Serve, we constantly add and _remove_ named actors.

**Comment by jovany-wang** [comment]
> This is great! Can you handle the deletion case as well? In Ray Serve, we constantly add and _remove_ named actors. It would be great to remove entries from the cache or keep the cache from growing infinitely. Specifically, the case I want to make sure that will work is (in pseudocode):

Yes, will refine it.

File: src/ray/core_worker/actor_manager.cc [hunk 1]
```diff
@@ -147,6 +148,105 @@ void ActorManager::HandleActorStateNotification(const ActorID &actor_id,
...
+
+  // We didn't find the actor in named actor cache locally,
+  // fetch it from Gcs synchronously.
+  auto result = SyncFetchNamedActorFromGcs(name, caller_id, call_site, rpc_address);
+  if (result.second.ok()) {
+    // Fetched the named actor from Gcs, let's cache it on local.
+    absl::MutexLock lock(&named_actor_cache_mutex_);
+    RAY_IGNORE_EXPR(actor_name_to_ids_cache_.emplace(name, result.first->GetActorID()));
+  }
+  if (result.second.ok()) {
```
Comment (rkooo567) [inline src/ray/core_worker/actor_manager.cc:176]: Both if has the same condition. Can we merge them to the same block?

```cpp
 if (result.second.ok()) {
    // Fetched the named actor from Gcs, let's cache it on local.
    absl::MutexLock lock(&named_actor_cache_mutex_);
    RAY_IGNORE_EXPR(actor_name_to_ids_cache_.emplace(name, result.first->GetActorID()));	    RAY_LOG(DEBUG) << "Success to cache the named actor on local with actor_name=" << name
                   << " and actor_id=" << result.first;
  }
```

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
@jovany-wang Please address the comment and fix a bug!

**Comment by edoakes** [comment]
@jovany-wang do you still plan to merge this?

**Comment by jovany-wang** [comment]
> @jovany-wang do you still plan to merge this?

I'm not going to merge this. 
After we have a better implementation, I'll create a new PR. see #13854

**Pull request closed without merging**